week_1_1.py

```diff
@@ -0,0 +1,154 @@
+"""
+
+TASK 1
+
+Write a class to represent a Cash Register.
+You class should keep the state of price total and purchased items
+
+Below is a starter code:
+------------------------
+1. you can add new variables and function if you want to
+2. you will NEED to add accepted method parameters where required.
+For example, method add_item probably accepts some kind of an item?..
+3. You will need to write some examples of how your code works.
+
+"""
+
+
+class CashRegister:
+
+    def __init__(self):
+
+        self.total_items = []
+        self.total_price = 0
+        self.discount = 0
+        self.total_items_details = []
+
+    def add_item(self, item, price):
+
+        # Adding item details dict to list.
+        # If that item has already been bought. i.e.,
+        # the customer is buying multiple of the same item,
+        # instead, increase the number bought by one and the price by one unit
+        # rather than having duplicates in the total_items_details list.
+        # I have kept duplicates in total_items.
+
+        if item not in self.total_items:
+            new_item = {
+                'item': item,
+                'no.': 1,
+                'price': price}
+            self.total_items_details.append(new_item)
+        elif item in self.total_items:
+            for i in self.total_items_details:
+                if i['item'] == item:
+                    i['no.'] += 1
+                    i['price'] += price
+
+        self.total_items.append(item)
+        self.total_price += price
+
+    def remove_item(self, item):
+
+        # Removing the one instance of the item from total_items_details
+        # if it indeed has already been 'checked out'.
+        # Removes the total price of that item from the total price.
+        # Removes all instances of the item from total_items.
+
+        if item in self.total_items:
+            for i in self.total_items_details:
+                if i['item'] == item:
+                    self.total_price -= i['price']
+                    self.total_items_details.remove(i)
+
+        while item in self.total_items:
+            self.total_items.remove(item)
+
+    def set_discount(self, discount):
+        self.discount = discount
+
+    def apply_discount(self):
+
+        # Changes all prices in self_items_details to the discount price and
+        # applies the discount to the whole of the total_price.
+        # And prints confirmation message.
+
+        for i in self.total_items_details:
+            i['price'] -= i['price'] * (self.discount/100)
+
+        self.total_price -= self.total_price * (self.discount/100)
```

Comment:
You have to update `self.total_price` in a couple of different places with different logic. You've also got two distinct things going on in this method updating the prices of individual items and calculating a new total (assuming that the old one was based on the sum of the individual items' values before the discount was applied).

I think it would be easier to follow and maintain if instead there was a separate method (perhaps a `@property`?) that would return `sum([item['price'] for item in self.total_items_details)`.